### ISSUE TITLE
Embedded XForms Server for mobile and offline clients

### ISSUE THREAD

**Issue by ebruchez** [orbeon/orbeon-forms#1221]
## Status

The purpose of this RFE is to gather ideas and track progress of an Orbeon Labs investigation project. The purpose of the project is to make the Orbeon Forms forms engine available on clients, including online and offline web browsers and mobile applications.

As of October 2016, there has been some progress on this project, including:
- conversion of our DOM library to Scala
- build modularization
- experimentation with XPath 2 in the browser

We expect to continue progress on this project incrementally.

## Subprojects, dependencies, and related projects

- [x] XPath 2 processor/API with JVM and Scala.js support #2127
- [x] XPath: replace reflective calls with native functions #2214
- [x] Don't use dom4j XPath APIs #2164
- [x] Migrate to a minimal version of dom4j #2165
- [x] Complete conversion of XForms engine to Scala #2825
  - did all that's needed, with some leftovers

These are related subprojects

- Requirements for an offline iOS/Android app #1545
- Mobile/offline Form Runner UI

## Historical note

With Orbeon Forms 3.7 (back in 2009!), we had an implementation which wasn't bad in concept, but was too limited and based on Google Gears (the precursor to HTML5 local storage among other things). Due to those limitations, Google abandoning Gears and lack of interest from customers at that time, we decided not to maintain it anymore starting Orbeon Forms 3.8. See:
- [XForms - Offline Mode Documentation](http://wiki.orbeon.com/forms/doc/developer-guide/offline-mode-documentation)
- [XForms - Offline Mode Internals](http://wiki.orbeon.com/forms/doc/contributor-guide/offline-mode-internals)
- [How can I create forms which users can fill out offline?](https://github.com/orbeon/orbeon-forms/wiki/FAQ-~-Orbeon-Form-Builder-and-Orbeon-Form-Runner#how-can-i-create-forms-which-users-can-fill-out-offline)
## Rationale

The idea is to bring our forms engine to Web apps and "native" mobile apps. This would serve multiple purposes:
1. supporting offline form management and data entry with a Web app
2. supporting offline form management and data entry from with a native phone/tablet app
3. generally having more of the form processing handled on the client, as client capabilities improve

The general idea is to make the code which currently runs on the server run on the client, whether within a Web browser directly or as a native application on a mobile device.

We need to abstract and modularize the XForms engine to make it as small as possible and as independent as possible from other things including the servlet API.

The user-interface part would remain browser technology-based. We don't want to change this at all at first and we wants to leverage either an actual Web browser or a Web view within a native app.

One components that we need to have is some kind of abstraction for the client/server communication. Right now that communication is handled via Ajax in the JavaScript code and we have the XForms server on the server. What we would like to do here is allowing the option to have something local like at simple queue. That doesn't seem to be extremely hard to do.
### General considerations
- We plan to compile our code to JavaScript with [Scala.js](https://github.com/lampepfl/scala-js).
  - it is no longer experimental as of 2015
  - it can run in a [`WKWebView`](https://developer.apple.com/library/prerelease/ios/documentation/WebKit/Reference/WKWebView_Ref/index.html) as of iOS 8
- We have a mix of Java and Scala code and we plan to migrate the Java code that is needed to Scala. (#2825)
- We are ruling out the use of GWT, which is a tool chain made obsolete (for us) by Scala.js.
### Web app

For a web application we have to compile our code to JavaScript. For offline support we can use modern browser APIs.
### Native iOS app

The current plan is to use JavaScript as well. In the future, and if it matures and gains the appropriate bindings, [Scala Native](https://github.com/scala-native/scala-native) could become an option as well.
### HIstorical considerations

We also considered these options, but have opted not to use them:
1. [Scala -> scalac -> IKVM -> Xamarin.iOS -> ARM x86 assembly](https://github.com/samskivert/ios-scala-demo)
   - iKVM is stable
   - MonoTouch is stable (?)
   - uses optimizing compiler (I think)
   - drawback: painful toolchain
   - benefit: might be the best performance possible as of 2013
   - opens the door of Microsoft phones/tablets as well
2. [Avian VM](http://oss.readytalk.com/avian/)
   - promising (2013-08-30 update: [Scala runs fully](https://groups.google.com/d/msg/scala-internals/gA3syROZPLA/o5STy4QYnDoJ))
   - JIT/AOT compilers and GC performance still poor

Abandoned projects, for reference:
1. [RoboVM](http://www.robovm.org/) compiles Java bytecode to native code (project/product is [dead](https://robovm.com/robovm-winding-down/))
2. [ScalaGWT](https://github.com/scalagwt/scalagwt-scala)
   - run in WebView
   - no JIT in WebView probably
3. [LLVM backend](http://greedy.github.io/scala-llvm/)
4. [J2ObjC](http://j2objc.org/) from Google
   - compiles Java source to Objective-C
   - see also [Going under the hood of Inbox](https://www.instapaper.com/read/543943804)
   - would not work for us almost half of our server-side code is in Scala
### Native Android application

Android already supports Java bytecode. This means we have two options:
1. Compile our Java/Scala code to Android.
2. Compile our Scala code to JavaScript as in the case of the browser and iOS.

We would need to use a WebView probably and then we would need to interact between the application and the WebView. On Android the JavaScript side of a WebView is able to call a Java object that has been exposed.
## Modularization effort

The Orbeon code needs to be modularized and abstracted. In particular:
- client-server connection
- temporary files
- form compilation vs. form execution
- HTTP client for submissions
- etc.
## Subsequent steps

Once a prototype is developed, we needs to think about the general architecture of the mobile application, including lifecycle, how to load forms and so on. Form Runner needs to be included, and local storage at least needs to be used.

**Comment by ebruchez** [comment]
## Steps for a prototype

### Strategy

As of 2015-01-13, the first approach we would like to try is Scala.js. This have benefits as a single compilation can be used:

- in actual web browsers
- in native apps via web views
- also allows us to master Scala.js, which we want to use more for Form Builder in particular

### First goals

We are trying to define reasonable chunks of work which lead us somewhere. First goals:

1. run *some* of our code in the browser
2. run an XForms hello world prototype in the browser

It is ok to hack to get to these goals. The purpose is to get familiar with the tool chain, potential issues, and learn more about what remaining work will be needed.

### Known challenges

- Java code/libraries don't compile with Scala.js right now
  - XPath 2 engine from Saxon
    - maybe hack in pseudo XPath engine (might need Saxon CE engine later)
  - Dom4j
    - remove dom4j dependency for static/dynamic/instances
    - need to decide how to store instances: browser DOM, anything else?
  - joda-time?
  - commons-(codec|io|lang3)?
  - our own Java code (which we can migrate)
- circumscribe/modularize/refactor
  - we only want to pull in a small subset of our code

### Things the client must not need

- XSLT 2 (later dynamic XBL might require a solution)
- XPL and processors
- eXist, iText, Flying Saucer
- most other XML processing code (MSV, XML Schema validator)
- Java libraries: httpclient, ehcache
- Java logging libraries (need some kind of wrapper)
- might not need HTML handlers at first at least
    - client would receive always same HTML "template" from server
    - all controls non-relevant
    - initial show does a diff with fully non-relevant tree (null)
    - so no full updates for now

### General architecture

- client would take
    - HTML page produced by server
    - serialized version of static state
        - must not depend on backing Dom4j `element`
        - could be XML or JSON or anything that can rebuild `ElementAnalysis` trees
        - this way complex XBL logic, XSLT etc. remains on server

**Comment by ebruchez** [comment]
![form-runner](https://cloud.githubusercontent.com/assets/105769/7644505/87101ae0-fa5a-11e4-85d9-d8cbacbc9d14.png)

**Comment by ebruchez** [comment]
[+1 from customer](https://mail.google.com/mail/u/0/#inbox/15633d19efe85edb)

**Comment by ebruchez** [comment]
[+1 from evaluator](https://mail.google.com/mail/u/0/#inbox/1579086be36f3e7c)

**Comment by ebruchez** [comment]
Immediate concrete next steps:

- [ ] create an XForms module in the build (we have progressed a lot on build modularization already)
- [ ] convert at least P1 classes to Scala (#2825)
- [ ] after that: determine next step for further modularizing XForms engine

**Comment by ebruchez** [comment]
[+1 from evaluator](https://mail.google.com/mail/u/0/#inbox/1582c30a1d316012), with a preference for a web app using browser storage, as native apps can be more challenging to deploy.

**Comment by ebruchez** [comment]
[+1 from customer](https://3.basecamp.com/3600924/buckets/1977373/messages/327513843#__recording_327551633)

**Comment by ebruchez** [comment]
[+1 from evaluator](https://mail.google.com/mail/u/0/#inbox/15cd063670e9a284)

**Comment by ebruchez** [comment]
[+1 from user](http://discuss.orbeon.com/Offline-forms-acces-td4663941.html)

**Comment by siebertm** [comment]
We as a customer would also +1 that. Just the form runner, right?

**Comment by avernet** [comment]
@siebertm Correct: this feature is specifically for forms created with Form Builder and running in Form Runner.

**Comment by ebruchez** [comment]
[+1 from user on StackOverflow](https://stackoverflow.com/questions/56494222/mobile-version-of-orbeon-forms)

**Comment by ebruchez** [comment]
This feature will also help us write more responsive UI for Form Builder. We'd like, for example, Form Builder dialogs showing and navigating faster and reducing the number of roundtrips to the server.

**Comment by ebruchez** [comment]
The proxy portlet preferences could also benefit.

**Comment by FOC-framework** [comment]
Hello, Is there any news regarding offline form runner? It is really a requirement from our business people. On the other side how does the Form builder store the form layout. Is it a JSON or XML file? And is the response also an XML JSON file? In which case one would imagine coding a new renderer for mobiles and Flutter can be an option for cross platform compatibility.

**Comment by avernet** [comment]
@FOC-framework Offline support is still something we'd like to do, but just not actively working at this point in time. Form definitions are stored in an XML format, which essentially looks like HTML + XForms + some Orbeon Forms specific elements and attributes. You can see what it looks like by going, in Form Builder, to Advanced / Edit Source. For services called by the forms, Orbeon Forms supports both XML and JSON.

**Comment by ebruchez** [comment]
Moving and reworking the steps into this separate comment:

## Proposal for next steps

### 1. Initial XForms engine extraction

Goals:
- must compile with scalac only (no Java code)
- must only use Scala code/libraries
- exclude XSLT, XPL, HTTP client, etc. (everything not core XForms)
- get a mostly extracted XForms engine
- get a list of remaining library dependencies if any
- get a list of further tasks that might come up
- if possible: do this as refactoring which can be merge on master branch

Non-goals:
- remove all Java dependencies
- compile with Scala.js

Steps:
- move things around until they compile ;)
- write interfaces & stubs where needed, hacking is ok

### 2. First compilation with Scala.js

Goals:
- reasonable subset of XForms engine compiles with Scala.js
  - XPath dependency analysis is explicitly excluded
  - use `DumbXPathDependencies`
- clarify next steps

Non-goals:
- code doesn't need to run or do anything useful yet
- having a functional XForms engine

Steps:
- if needed, stub out XPath engine, BUT could also hookup Darius XPath

### 3. Simple client-side Hello World

Goals:
- run Hello World in the browser

Non-goals:
- fully functional XForms engine (no full XPath support)
- process XForms source on the client

Steps:
- client/server infrastructure
  - serialize/deserialize static state
  - generate initial HTML with serialized states
  - hookup local Ajax server
- hookup XPath processor (Darius XPath)
- interface with DOM for the UI

### 4. Full XForms support

Goals:
- most XForms must work
- complete XPath support
  - XPath 2 processor/API with JVM and Scala.js support #2127
  - Java/Scala bridge for function calls without reflection 
  - typed values
  - hookup XPath function library
  - dependencies (`PathMap`) support
- complete XForms support that makes sense
  - submissions
  - TODO

Steps:
- TODO

### 6. Initial Form Runner support

Goals:
- some forms built with Form Runner must work
- basic support for taking a Form Runner form offline and sync back online

Steps:
- TODO

### 7. Full Form Runner support

Goals:
- Requirements for an offline iOS/Android app #1545

### 8. Option: demo iOS app

Goal:
- run the above in a simple iOS app
- must show basic XForms, interactive, in web view

Steps:
- TODO

**Comment by siebertm** [comment]
@ebruchez that's a large chunk of work and i know this is rather far out but if you're at the point you've got an MVP or so to test out, ping me! I really want to try this in our real use cases!

**Comment by ebruchez** [comment]
The initial refactoring requires separating the following:

- [ ] form runtime
- [ ] form compilation
- [ ] submission module
- [ ] client/server communication
- [ ] TODO

**Comment by ebruchez** [comment]
- [ ] eventually: two separate subprojects, for example `xforms-runtime` and `xforms-compiler`
- [ ] `XFormsContainingDocument` → Scala
- [ ] static state: separare construction from data structures

**Comment by ebruchez** [comment]
We will need the handlers offline to support instantiating new forms offline. Right now, this works as follows:

- take the `SAXStore` with the template
- replay this through the handlers system'
- the output is SAX events
- that's then serialized

So we need a way to do this when offline as well. We probably need to have `SAXStore` and some of the SAX API for this.

**Comment by ebruchez** [comment]
Splitting this into modules/projects is not that trivial. We not only have the runtime/compile-time axis, but the "thin client" (what we have now)/full client axis, and the JVM/JS/shared axis.

For example we don't want the "thin client" to pull down the full client. Scala.js prunes the call tree but it's not perfect at doing that, probably, so it's better if we have a clean separation of the "full client" module, which would contain mainly the XForms runtime.

**Comment by ebruchez** [comment]
How static state construction works:

- `XFormsStaticStateImpl.restore` / `.createFromStaticStateBits` / `.createFromDocument` (tests)
- `new PartAnalysisImpl`
- `topLevelPart.analyze()`
    - `new RootControl`
    - recursive call to `build()`
    - final hooking up

**Comment by ebruchez** [comment]
Client-server communication abstraction: 

- `AjaxClient` must be abstracted or call to a common interface
    - currently this depends on JS
- we probably keep the event queue in all cases
- we need a shared event representation ideally (so later, the client-server protocol can use those too)
    - currently `AjaxEvent` depends on JS
- `ClientServer` still extracts information from `Element`

**Comment by ebruchez** [comment]
So for the client-server, we now have:

- `ClientServerChannel`
- `WireAjaxEvent`
- The response is still XML and will have to be updated as well.

For `XFormsModelSubmission`:

- `RegularSubmission` uses `Connection`, which uses `PropertiesApacheHttpClient` and `InternalHttpClient`
- neither client is appropriate when offline
- but when running fully in JS, something like a `FetchHttpClient` would be good, although it would have to be async
    - See "Switch the actions implementation to use async submissions #4077"
- when fully offline, we should have an `OfflineHttpClient`, configurable to call back JavaScript functions

So probably that we don't need to modify much `XFormsModelSubmission`. One approach:

- make `Connection`'s choice of `HttpClient` pluggable
    - on the JVM, use `PropertiesApacheHttpClient` and `InternalHttpClient`
    - on JavaScript, use `OfflineHttpClient`
- remove a few dependencies
    - `org.apache.http.client.CookieStore`/`Cookie`
    - etc.

Another approach is to make `Submission`s pluggable, and on JavaScript:
    - remove `RegularSubmission`, and possibly others
    - instead, use a new `OfflineSubmission`

The second approach might be easier as it's not pulling in `Connection`.

**Comment by siebertm** [comment]
My 2 cents on this:

> when fully offline, we should have an OfflineHttpClient, configurable to call back JavaScript functions

Why not normal `fetch`/`XHR` which are then caught in a `ServiceWorker`? This generates less special cases. 

Depending whether this must support IE, of course. But, even we (working with german hospitals) are phasing out IE support...

**Comment by ebruchez** [comment]
@siebertm Thanks for the feedback.

> Why not normal `fetch`/`XHR` which are then caught in a `ServiceWorker`? This generates less special cases.

The main reason for a JavaScript callback API is embedding in native mobile applications which can then provide their own persistence and service support. It's also very easy to implement on both sides.

For regular online/offline mode in a browser then yes, Fetch support is needed.

> Depending whether this must support IE, of course. But, even we (working with german hospitals) are phasing out IE support...

Agreed, and IE support is not a requirement at this point.

**Comment by ebruchez** [comment]
For the record, we have now made `Connection` implement `ConnectionTrait`, with two implementations (JVM/JS). The JS implementation is not done yet.

More work:

- [ ] `CacheableSubmission` depends on `XFormsServerSharedInstancesCache`
- [ ] provide implementation of `Connection` based on API callbacks and/or Fetch

**Comment by ebruchez** [comment]
Which parts of the static analysis must not be present in the `ElementAnalysis` tree?

- XPath analysis
    - this is done lazily (I think) in `SimpleElementAnalysis`
   - we *could* do it on the client, I guess, but we shouldn't need to
- calls to `ComponentControl.setConcreteBinding`, which uses `XBLBindingBuilder`
    - this is the big one!
    - `ShadowChildrenBuilder`, which is implemented by `ComponentControl`, and uses `binding.compactShadowTree`

So we should probably separate the "builders": `ChildrenBuilderTrait`, `ShadowChildrenBuilder`, etc.

**Comment by ebruchez** [comment]
So now we are refactoring out all the building of the tree. This will take care of `XBLBindingBuilder` and others.

However we need to solve the question of how to pass in the XPath analysis information. We do this after we have built the tree:

- `rootControlAnalysis.analyzeXPath() // Analyze root control XPath first as nested models might ask for its context`
- `analyzeModelsXPath() // Analyze all models XPath`
- `analyzeControlsXPath() // Analyze controls XPath`

For deserialization, the information should be passed to the constructor. So maybe the constructor of the classes should have the XPath analysis information as `var`s set to `None`, but then the `var`s are updated before serialization. Upon deserialization, the proper information can be passed.

**Comment by ebruchez** [comment]
Next step is to move towards serialization/deserialization. This includes:

- `ElementAnalysis` hierarchy
- `Scope`
- `Metadata`: must include `Mark`, `NamespaceMappings` (but no `IdGenerator`, XBL index, etc.)

We can use `case` classes with `var`s, or semi-automatic, or manual serialization.

**Comment by ebruchez** [comment]
Remaining tasks before we can rebase on the pe branch:

- [ ] fix all tests on offline branch
- [ ] resolve Saxon HE contributions
- [ ] clean existing commit log (for example  "xxx" in logs)
- [ ] remove `println`s
- [ ] check `freeTransientState` calls/frees everything
- [ ] "TODO: pass a logger?"
- [ ] 

On master (CE) branch:

- [ ] decide which code/subprojects don't belong
    - [ ] remove `xformsOffline`
    - [ ] remove `formRunnerOffline`

P2 cleanup tasks for Saxon:

- [ ] `LoofLifter`
- [ ] `JsonParser`
- [ ] qt test suite
- [ ] port Saxon patches

P3:

- [ ] `XFormsServerSharedInstancesCache` support for `timeToLive`
- [ ] "// 2020-12-23: If the provider is not configured, return `false` (for offline FIXME)."
- [ ] `AsynchronousSubmissionManager`
- [ ] check `XFormsStateManager`
- [ ] new tests
    - [ ] `SimplePath`
    - [ ] `NodeWrapper`
    - [ ] `orbeonDomToTinyTree`
- [ ] FIX: fall through in `PathMap`

Saxon P3:

- [ ] don't use own `Trie`, just use Scala collections
- [ ] some uses of Scala `Iterator` vs. `ju.Iterator`
- [ ] use more Scala-like enums
- [ ] `BeanProperty`
    - 2020-10-09: 406 occurrences

**Comment by ebruchez** [comment]
For 2021.1 this is not a full-fledged feature. Our plan is to enable, possibly optionally, the "Test Offline" feature from Form Builder. This feature allows running a form in test mode using the JavaScript form runtime.

**Comment by ebruchez** [comment]
- [ ] move P2/P3 to follow-up issue
- [ ] doc
